File: LeetCodeExample.Test/GoogleQuestion/Acres/_00046_Permutations.cs
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Linq;

namespace LeetCodeExample.Test.Google
{
    /// <summary>
    // Given an array nums of distinct integers, return all the possible permutations.You can return the answer in any order.
    /// </summary>
    public class _00046_Permutations
    {
        public IList<IList<int>> Permute(int[] nums)
        {
            rtn = new List<IList<int>>();
            List<int> coll = new List<int>();
            Perm(nums, coll);
            return rtn;
        }

        IList<IList<int>> rtn;
        void Perm(int[] nums, List<int> coll)
        {
            // Base case
            if (coll.Count == nums.Length)
            {
                List<int> copy = new List<int>(coll);
                rtn.Add(copy);
                return;
            }

            for (int i = 0; i < nums.Length; i++)
            {
                // Make sure this index is not in coll
                if (coll.Contains(nums[i]))
                    continue;

                coll.Add(nums[i]);
                Perm(nums, coll);
                // Backtrack
                coll.RemoveAt(coll.Count - 1);
            }
        }
    }
}

File: LeetCodeExample.Test/Combinations/_00046_Permutations.py
# Given an array nums of distinct integers, return all the possible 
# permutations
# . You can return the answer in any order.

from typing import List
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        def rec(lst: list, seen: set):
            if len(lst) == len(nums):
                ans.append(list(lst))
                return
            for num in nums:
                if num in seen:
                    continue
                seen.add(num)
                lst.append(num)
                rec(lst, seen)
                seen.remove(num)
                lst.pop()
        
        rec([], set())
        return ans